org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java

```diff
@@ -596,6 +620,93 @@ protected Compiler newCompiler() {
 	return newCompiler;
 }
 
+private CompilerConfiguration prepareCompilerConfiguration(CompilerOptions options) {
+	try {
+		List<String> annotationProcessorPaths = new ArrayList<>();
+		List<String> generatedSourcePaths = new ArrayList<>();
+		boolean isTest = this.compilationGroup == CompilationGroup.TEST;
+		if (this.javaBuilder.participants != null) {
+			for (CompilationParticipant participant : this.javaBuilder.participants) {
+				if (participant.isAnnotationProcessor()) {
+					String[] paths = participant.getAnnotationProcessorPaths(this.javaBuilder.javaProject, isTest);
+					if (paths != null) {
+						annotationProcessorPaths.addAll(Arrays.asList(paths));
+					}
+					String[] generatedSrc = participant.getGeneratedSourcePaths(this.javaBuilder.javaProject, isTest);
+					if (generatedSrc != null) {
+						generatedSourcePaths.addAll(Arrays.asList(generatedSrc));
+					}
+				}
+			}
+		}
+
+		ClasspathLocation[] classpathLocations = this.nameEnvironment.binaryLocations;
+		Set<String> classpaths = new LinkedHashSet<>();
+		Set<String> modulepaths = new LinkedHashSet<>();
+		for (ClasspathLocation location : classpathLocations) {
+			if (location instanceof ClasspathDirectory cpDirectory) {
+				IPath binaryLocation = cpDirectory.binaryFolder.getLocation();
+				if (binaryLocation == null) {
```

Comment:
Such null check does not allow compilers to be compliant to handle external files. Better just collect the IFolders instead of string path such that external folders are not lost.